parcels/particleset/particlesetsoa.py

```diff
@@ -427,6 +478,38 @@ def to_dict(self, pfile, time, deleted_only=False):
         return self._collection.toDictionary(pfile=pfile, time=time,
                                              deleted_only=deleted_only)
 
+    def compute_neighbor_tree(self, time, dt):
+        active_mask = self.active_particles_mask(time, dt)
+
+        self._values = np.vstack((
+            self._collection.data['lat'],
+            self._collection.data['lon'],
+            self._collection.data['depth'],
+        ))
+        if self._dirty_neighbor:
+            self._neighbor_tree.rebuild(self._values, active_mask=active_mask)
+            self._dirty_neighbor = False
+            self._neighbor_time = time
+        else:
+            self._neighbor_tree.update_values(self._values, new_active_mask=active_mask)
+
+    def neighbors_by_index(self, particle_idx):
+        neighbor_idx, distances = self._neighbor_tree.find_neighbors_by_idx(
+            particle_idx)
+        neighbor_idx = self._active_particle_idx[neighbor_idx]
+        mask = (neighbor_idx != particle_idx)
+        neighbor_idx = neighbor_idx[mask]
+        if 'surf_dist' in self._collection.data:
+            self._collection.data['surf_dist'][neighbor_idx] = distances[0, mask]
+            self._collection.data['depth_dist'][neighbor_idx] = distances[1, mask]
+        return ParticleCollectionIterableSOA(self._collection, subset=neighbor_idx)
+
+    def neighbors_by_coor(self, coor):
+        # TODO: fix if necessary
+        neighbor_idx = self._neighbor_tree.find_neighbors_by_coor(coor)
+        neighbor_ids = self._collection.data['id'][neighbor_idx]
+        return neighbor_ids
```

Comment:
yeap, nice function.

The more I see that, the more I see that adapting it for other structures may entail a new super-class here ... anyway: nice function.